homeassistant/helpers/script.py

```diff
@@ -767,11 +783,19 @@ async def async_run(
                 self._log("Maximum number of runs exceeded", level=logging.WARNING)
                 return
 
+        # If this is a top level Script then make a copy of the variables in case they
+        # are read-only, but more importantly, so as not to leak any variables created
+        # during the run back to the caller.
+        if self._top_level:
+            variables = dict(variables) if variables is not None else {}
+
         if self.script_mode != SCRIPT_MODE_QUEUED:
             cls = _ScriptRun
         else:
             cls = _QueuedScriptRun
-        run = cls(self._hass, self, variables, context, self._log_exceptions)
+        run = cls(
+            self._hass, self, cast(dict, variables), context, self._log_exceptions
```

Comment:
If the script is not `self._top_level` then `variables` is never `None`?

Reply:
Correct.

A _non_ top-level `Script` object is only created and used by another `Script` object (in the implementation of the `repeat` or `choose` actions) and is never called by anything but the "parent" `Script` object, ultimately up to the "top level" `Script` object, which, given this new code, guarantees its children will always get a `dict` object and not `None`. That's why I changed the typing info for the variables passed into `_ScriptRun.async_run()` to be _not_ optional.

The `dict` copy, created by the parent, is used throughout the entire "run" so that variables created during the run, possibly in a "child" (aka "sub-sequence") can exist when the child is done. This part isn't important yet, because it only applies to the `repeat` action so far, and that effectively creates a "stack" of `repeat` variables, because `repeat`'s inside other `repeat`'s "hide" values from outer the outer loops (although I may at some point make the ones from the outer loops accessible somehow. But that's for another time.) When this becomes more important is with the `wait_for_trigger` action. In that case the `wait_trigger` variable needs to exist _after_ the `wait_trigger` executed, in time, regardless of what "level" it ran in. But, again, as long as I'm fixing the variable handling, I might as do it right now as later.

If it helps to understand, a top-level `Script` object will create "children" `Script` objects for `repeat` or `choose` actions. The "run" calling sequence is then:
```python
# --------------- top-level ---------------     ------------------ child -------------------
Script.async_run() -> _ScriptRun.async_run() -> Script.async_run() -> _ScriptRun.async_run()
```
Only that very first call might get `None`, or a read-only `MappingProxyType` object. Everything after that will use the writable `dict` copy created here.